### DISCUSSION TITLE
ARM/x64 macOS universal binary triplet

### DISCUSSION THREAD

**Discussion by Be-ing** [microsoft/vcpkg#19454] [Ideas]
****

**Is your feature request related to a problem? Please describe.**
Apple has switched CPU architectures again, this time to ARM.

**Proposed solution**
Add a new triplet arm64-x64-osx (or some similar name) for ARM/x64 universal binaries

**Describe alternatives you've considered**
Make separate builds for x64-osx and arm64-osx

**Additional context**
https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_ARCHITECTURES.html

**Comment by JackBoosY** [comment]
Yes, we are planing to do that.

**Comment by JackBoosY** [comment]
cc @strega-nil-ms

**Comment by strega-nil-ms** [comment]
We've talked about how to do this, but haven't had the chance to design a good system for this yet. We really need to spend some serious time on OSX in this.

**Reply by Be-ing** [reply]
Has there been any progress towards this?

**Comment by kenr** [comment]
Do you have a triplet name in mind for fat binaries?

**Comment by kelteseth** [comment]
Is there any workaround for this until we get a new triplet? @strega-nil-ms

**Reply by Nazgul37** [reply]
@kelteseth Have you been able to find a way to do this?

**Reply by kelteseth** [reply]
I have not. I create two separate executables for now...

**Reply by Be-ing** [reply]
You can make two separate builds and then combine them with `lipo`.

**Comment by Nazgul37** [comment]
Has there been any update on the best way to do this at current? @JackBoosY

**Reply by JackBoosY** [reply]
We are always waiting for additional budget.

**Reply by Nazgul37** [reply]
Oh, I see. Are you aware of any reliable workarounds that could be used in the meantime?

**Reply by JackBoosY** [reply]
To make arm64-osx as the next official triplet, we should add new machines.

**Reply by kenr** [reply]
Any updates? We use vcpkg a lot in our organization and our customers are very interested in using M1 versions of our software.

**Reply by konstantinkopka** [reply]
@JackBoosY I am also interested in having support for universal binaries on macOS, as another project I'd like to use needs that for Apple Silicon support (https://github.com/TrenchBroom/TrenchBroom/issues/4145#issuecomment-1320861387)

Any update on whether or not this might be possible would be very much appreciated :)

**Reply by JackBoosY** [reply]
@BillyONeal Do we have any plans for this recently?

**Reply by BillyONeal** [reply]
We are still in a holding pattern right now on the hardware; we met with the folks who own that stuff back in December; they asked us to meet again in the new year so maybe we'll have it soon.

**Reply by Dariusz1989** [reply]
Any update ?

**Reply by photex** [reply]
I've started advocating for vcpkg over Conan at work, part of the reason is because the triplet model is a lot more promising than Conans flat out refusal to support or acknowledge universal binaries.

Has anyone attempted their own triplet while waiting for the official solution?

Any updates from Microsoft would be really helpful.

**Reply by kelteseth** [reply]
@photex Incompile the vcpkg dependencies twice and then combine them with lipo, see my python script https://gitlab.com/kelteseth/ScreenPlay/-/blob/master/Tools/macos_lipo.py

**Reply by photex** [reply]
I just tested a custom triplet 'universal-osx.cmake' which was a copy of arm64-osx.cmake from the community triplets. 

I changed the architectures line to:
`set(VCPKG_OSX_ARCHITECTURES "arm64;x86_64")`

I built a universal binary of SDL2 and linked with it.

**Reply by BillyONeal** [reply]
> Any updates from Microsoft would be really helpful.

arm64-osx may be in the cards if we can finally get a way to get the machines we have into CI. I don't expect a universal triplet anytime soon because there are too many build systems that themselves don't understand this. As a result I suspect we and Conan are in the same place in recommending separate builds and lipo afterwards.

Note that several of vcpkg's triplet knobs like VCPKG_CXX_FLAGS do not make sense when there are literally two separate compiles happening, so even if we add such a thing it's going to be very much it's own snowflake.

**Reply by photex** [reply]
How does or would the lisp method work with manifest mode? It’s just not part of the build then I guess.

**Comment by faaxm** [comment]
Is there some work that has already been started in that direction and that could be picked up by the OSS community?

I have been using the process outlined in the blog-post below, but would be nice to have official support.
https://www.f-ax.de/dev/2022/11/09/how-to-use-vcpkg-with-universal-binaries-on-macos/

**Comment by asad26** [comment]
Hi @JackBoosY, @BillyONeal

Is there any update on supporting universal macOS triplet? Thanks.

**Reply by kelteseth** [reply]
From what I've gathered, and personal experience is that it is a build system issue. When all of your dependencies are using CMake, you would be fine using a osx-unversal triplet and set the CMake config to:

``` cmake
set(VCPKG_TARGET_ARCHITECTURE arm64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)

set(VCPKG_CMAKE_SYSTEM_NAME Darwin)

# THIS WILL _NOT_ WORK FOR NONE CMAKE LIBS:
set(VCPKG_OSX_ARCHITECTURES "arm64;x86_64")
```

This will **not for work any lib that does not use CMake**, aka most libraries. This means you must compile it twice and use the script from https://github.com/microsoft/vcpkg/discussions/19454#discussioncomment-6667878

**Comment by BardGysler** [comment]
Hi, any update about this?

**Reply by autoantwort** [reply]
No

**Comment by sharadhr** [comment]
Some experiments: AppleClang `clang` and `clang++` support the `-arch` flag, which is (partially) how [CMake implements `CMAKE_OSX_ARCHITECTURES`](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Source/cmLocalGenerator.cxx#L1987).

Additionally, the `-isysroot` flag is [set](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/Platform/Apple-Clang.cmake#L12) with [`CMAKE_OSX_SYSROOT`](https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_SYSROOT.html) and [used in the compiler invocation](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Source/cmLocalGenerator.cxx#L1997) as well.

Of course, vcpkg straightforwardly supports universal CMake dependencies by forwarding `VCPKG_OSX_ARCHITECTURES` to `CMAKE_OSX_ARCHITECTURES`. For other build systems, I believe the best way forward would be to use CMake's built-in infrastructure under script mode (like vcpkg already does with GNU Autotools, for instance) to prepare `CFLAGS`, `CPPFLAGS`, `CXXFLAGS`, `LDFLAGS`, and `LIBS` in the configure-make invocations; the end result should have, at least, `-arch x86_64 -arch arm64` in `C[XX]FLAGS` and `LDFLAGS`. 

Some vcpkg portfile function implementations will have to be modified to properly support a `universal-osx` triplet.

**Comment by VioletGiraffe** [comment]
Is there an open issue for universal binary support? Clearly there has to be, but I can't find it.
